### PR TITLE
Add support for docker's client's --add-host command

### PR DESCRIPTION
**Pull request by sampwing**
Updated docker-py to the most recent version ( 0.7.1), which allows assigning values to extra_hosts (similar to ---add-host.)  Updated fig to allow passing in these parameters and added tests.

Implementation requested in the following issues:
#597 
#754

### CODE REVIEW

**Comment by sampwing** [comment]
Kinda stumped on this I test locally (script/test) and I don't get the two failures wercker ran into.  Currently running:
Docker version 1.4.1, build 5bc2ff8

**Comment by dnephin** [comment]
@sampwing wercker runs Docker 1.2.1 I believe https://github.com/wercker/wercker-labs-docker

I'm not sure if we can convince them to upgrade to latest, or if we'd have to create a custom box to run the tests.

**Comment by sampwing** [comment]
@dnephin I could look into creating a box if that could push this along - 1.2 is a bit out of date, fig seems to be recommending 1.4 on irc

**Comment by aanand** [comment]
I had a quick go yesterday at creating a custom Wercker box with Docker installed, but it failed with an error:

> Docker found, this is not supported in the default wercker environment. See for more information: http://devcenter.wercker.com/articles/docker/

This suggests that you're not allowed to build a custom box with Docker in it. I've asked them about it on Twitter - we'll see.

**Comment by thaJeztah** [comment]
@aanand given that fig/compose is part of docker, don't you have access to the build servers a Docker itself is using?

**Comment by tdesvenain** [comment]
Hi, what is the review status of this pull request be merged ? What is the next step ? I can help.
Thanks

**Comment by dnephin** [comment]
It needs a rebase and squash, but it's also blocked on #886

**Comment by ch3lo** [comment]
+1

File: fig/service.py [hunk 1]
```diff
@@ -629,6 +634,25 @@ def split_port(port):
...
+        extra_hosts_config = [extra_hosts_config]
+
+    extra_hosts_dict = {}
+    for extra_hosts_line in extra_hosts_config:
+        if isinstance(extra_hosts_line, dict):
+            # already interpreted as a dict (depends on pyyaml version)
+            extra_hosts_dict.update(extra_hosts_line)
+        else:
+            # not already interpreted as a dict
+            extra_hosts_dict.update(extra_hosts_line.split(':'))
```
Comment (ddunwoody) [inline fig/service.py]: I believe line 651 should read:

    extra_hosts_dict.update([extra_hosts_line.split(':')])

Comment (tdesvenain) [inline fig/service.py]: Actually this introduces a bug

I just commited a fix + some unit tests, could you review them please ?
https://github.com/tdesvenain/fig/commit/3e6c7cd5a49a52f68184a3d784066d00ab377c1d

2015-02-25 13:27 GMT+01:00 ddunwoody <notifications@github.com>:

> In fig/service.py
> <https://github.com/docker/fig/pull/848#discussion_r25335828>:
>
> > @@ -629,6 +634,25 @@ def split_port(port):
> >      return internal_port, (external_ip, external_port or None)
> >
> >
> > +def build_extra_hosts(extra_hosts_config):
> > +    if extra_hosts_config is None:
> > +        return None
> > +
> > +    if isinstance(extra_hosts_config, basestring):
> > +        extra_hosts_config = [extra_hosts_config]
> > +
> > +    extra_hosts_dict = {}
> > +    for extra_hosts_line in extra_hosts_config:
> > +        if isinstance(extra_hosts_line, dict):
> > +            # already interpreted as a dict (depends on pyyaml version)
> > +            extra_hosts_dict.update(extra_hosts_line)
> > +        else:
> > +            # not already interpreted as a dict
> > +            extra_hosts_dict.update(extra_hosts_line.split(':'))
>
> I believe line 651 should read:
>
> extra_hosts_dict.update([extra_hosts_line.split(':')])
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/fig/pull/848/files#r25335828>.
>



-- 
Thomas Desvenain

Téléphone : 09 51 37 35 18

**Comment by skrassiev** [comment]
hope this is the right way for +1

**Comment by albers** [comment]
+1

**Comment by phylake** [comment]
+1

**Comment by c0deaddict** [comment]
+1

**Comment by bmilesp** [comment]
+1

**Comment by aanand** [comment]
Needs a rebase. docker-py changes can probably go.

**Comment by slashmili** [comment]
+1

**Comment by jelmstrom** [comment]
+1

**Comment by tristan0x** [comment]
+1

**Comment by aanand** [comment]
Superseded by https://github.com/docker/compose/pull/1158

**Pull request closed without merging**